### PR TITLE
DEVC support

### LINKED ISSUE #909: DEVC Package / "development class" (missing parts)
For the package concept to work fully also in regards to package interfaces and package checks there are still some parts missing. Currently packages are created on clone utilizing the folder strategy defined in .abapgit.xml . Package attributes are not included in the abapGit repository. This includes:
- Package type (main package, structure package, development package)
- Package encapsulation flag
- Short description
- Application component
- Package interface usages and error severity

To version these attributes DEVC must be included in serialization and deserialization. Some stuff that might complicate things:
- The "local" packages where a repository is installed in can be named differently than the DEVC objects in the repo. A mapping must be used.
- The package concept has been changed a number of times within the release range that abapGit supports (like client/server checks, semantic interfaces, ...).
- Package hierarchies have restrictions. I. e. a main package cannot contain development objects other than subpackages and package interfaces.

In a second step package interfaces might be used as a _weak_ package / dependency management concept across abapGit repositories.

More thoughts can be found in today's slack discussion.

### PR DESCRIPTION
**Pull request by flaiker**
Fixes #909 

Not really finished but I thought I would open the PR to maybe get some feedback. This is kind of difficult to test :/

Some notes on the implementation:
- Object type DEVC is now also serialized, but only if `ADVANCED_PACKAGES` is switched on in .abapgit.xml (by default it is not)
- DEVC serialization includes all system and user independent attributes of a package (including main package, encapsulation and package interface usages)
- The existing LCL_SAP_PACKAGE functionality is still used for package creation, DEVC deserialization just changes the package attributes afterwards.
- Installing a repository to a different package also still works. I added an `ORIGINAL_PACKAGE` attribute to .abapgit.xml to specify how the files should be named. This needs to be set in order for different local package naming to work. In my testing $packages also worked fine (though they cannot have a component assigned, which will show up as a diff).
- LCL_FOLDER_LOGIC is still used for path to package conversion, though it was enhanced for deviating installation packages

I kind of trial-and-errored my way through the package-name-mapping problem, so some feedback there would be appreciated. I. e. I don't know if offline repositories or background mode are affected in any way, as I haven't used these features yet.

Test repository: https://github.com/flaiker/ztest

TODOS:
- [X] Add the new .abapgit.xml settings to the repo settings page (initialize `ORIGINAL_PACKAGE` if it's empty)
  - -> Settings were removed
- [X] Package deletion (should they be deleted? Are they currently? Seems kind of difficult. )
  - -> Package deletion is not supported for now
- [X] Fix DEVC jump (doesn't work if the installation package is different to the original one)
  - -> Implemented
- [ ] Backport to 702 (though I don't exactly know which APIs changed)

For removing the test repository after testing I used this procedure (hoping there's something less time consuming):
- Delete the package contents
- Delete empty packages (not including the ones that had objects in them before)
- Delete the lock entries in the transport task ignoring the warning (SE01)
- Delete the object catalog entries (SE03)
- Delete the empty packages
- Delete the new lock entries in the transport task
- Delete object catalog entries again

$packages don't require this procedure.

This is also the reason why package deletion seems difficult.

Open to any suggestions / feedback!

### CODE REVIEW

**Comment by larshp** [comment]
cool, will have a look tonight

**Comment by larshp** [comment]
Packages are currently not deleted by abapGit.

Regarding backport, I dont have a way of testing abapGit on 702, so I just push it and hope it works, if it doesnt somebody will hopefully create an issue.

Suggestion: is it possible in `lcl_tadir` to change the OBJ_NAME of all packages to "PACKAGE", this way repos using PREFIX will not contain the package name in any way. When serializing the package, IS_ITEM-DEVCLASS should contain the real package name, and when deserializing `iv_package` contains the real package name

**Comment by flaiker** [comment]
Good idea with the package name. I'll get on that.

I added the option so that existing repositories would not just switch to DEVC serialization on the next abapGit update but rather opt-in manually, while new ones would just have it enabled by default. But I suppose if it works well enough we can just drop the option all altogether.

One scenario I can think of where it would break something is when the target package hierarchy is not compatible. For example the top level repo package is a main package while the target package is a subpackage of a development package and can therefore not be a main package. Same with structure packages. Not too likely to happen though.

**Comment by flaiker** [comment]
Hmm, seems kind of like the warning for non-unique file names isn't just there for fun but rather files with the same name across different folders result in some problems. For example here:
https://github.com/flaiker/abapGit/blob/89b6e427c100b72d0d8e646da4bc44522ad9d4f7/src/zabapgit_objects.prog.abap#L490
`lcl_objects_files` will always get the first file with the name resulting on the others not being updated on pull. I tried to add path to the key of the read statement, but for that I kind of need to inject the installation package and `lcl_dot_abapgit` into the whole callstack so that you can call `lcl_folder_logic=>package_to_path`. Not sure if that is a good idea.

**Comment by larshp** [comment]
hmm, just hardcode to ignore "package.devc.xml" ?

**Comment by flaiker** [comment]
Ignore it where?

**Comment by larshp** [comment]
Hi,
I tried installing the code, but I do not get any warnings, did you fix the issue? Or?

File: src/zabapgit_dot_abapgit.prog.abap [hunk 1]
```diff
@@ -119,9 +119,9 @@ CLASS lcl_dot_abapgit IMPLEMENTATION.
     DATA: ls_data TYPE ty_dot_abapgit.
 
 
-    ls_data-master_language = sy-langu.
-    ls_data-starting_folder = '/'.
-    ls_data-folder_logic    = c_folder_logic-prefix.
+    ls_data-master_language   = sy-langu.
+    ls_data-starting_folder   = '/'.
+    ls_data-folder_logic      = c_folder_logic-prefix.
```
Comment (larshp) [inline src/zabapgit_dot_abapgit.prog.abap]: can this include be reset to original so it does not show up in the diff?

File: src/zabapgit_object_devc.prog.abap [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+
+  METHOD lif_object~compare_to_remote_version.
+    CREATE OBJECT ro_comparison_result TYPE lcl_comparison_null.
+  ENDMETHOD.
+
+  METHOD lif_object~delete.
+    DATA: li_package    TYPE REF TO if_package,
+          lv_changeable TYPE abap_bool.
+
+    RETURN ##TODO.
```
Comment (larshp) [inline src/zabapgit_object_devc.prog.abap]: just RETURN here, the packages cannot be deleted as part of object deletion, also packages are currently not deleted by abapGit

**Comment by larshp** [comment]
hmm, new approach, it seems like a mess renaming everything to PACKAGE, a lot of code assumes unique naming of objects.

Try in https://github.com/larshp/abapGit/blob/master/src/zabapgit_objects_impl.prog.abap#L516 for all DEVC, instead of using `is_name` just name all xml files for DEVC = `package.devc.xml`, this way we can keep the internal logic, and still serialize to a common filename.

And when finding the item name from the filename, we need to override "package" by running the folder logic to determine the internal name for the package

**Comment by flaiker** [comment]
Seems to work quite well for me now.

There are some edge cases, for example when deleting packages or pulling new packages where the folder logic seems to somehow lock the package, but it works fine after restarting abapGit. I also found some issues when moving existing objects to a different package, especially regarding dumps on "reset local" because of duplicates keys; I don't think it has anything to do with this PR however. FULL folder logic also seems to work fine.

**Comment by larshp** [comment]
cool, will test it sometime this week

**Comment by larshp** [comment]
Lets do it 🚢 Merging

Downport: we'll fix the issues if/when they show up

**Pull request merged**